src/transformStateWithClones.js

```diff
@@ -7,7 +7,38 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = state;
+
+  actions.forEach((action) => {
+    let nextState;
+
+    switch (action.type) {
+      case 'clear':
+        nextState = {};
+        break;
+      case 'addProperties':
+        nextState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        nextState = { ...currentState };
+
+        action.keysToRemove.forEach((key) => {
+          if (nextState.hasOwnProperty(key)) {
+            delete nextState[key];
+          }
+        });
+        break;
+      default:
+        nextState = { ...currentState };
+        break;
+    }
+
+    stateHistory.push(nextState);
+    currentState = nextState; // Update currentState for the next iteration
```

Comment:
Instead of pushing 'nextState' directly to 'stateHistory' and then assigning 'currentState' to 'nextState', consider doing this operation once after the switch statement has completed to avoid repetition.